### PR TITLE
Fix vibevoice TTS batched audio index

### PR DESCRIPTION
**Pull request by ebezzam**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48902) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48902)

**In short**, while benchmarking VibeVoice TTS for the [Open TTS Leaderboard](https://huggingface.co/spaces/hf-audio/open_tts_leaderboard), I noticed some issues with batched inference. Namely cross-talk between individual samples when one audio finishes generating earlier than another audio.

The PR includes the fix needed in `generation_vibevoice.py`, and Claude helped put together a test (that would fail in the original case).

### _Long AI description_ 👇 

`_decode_audio_latent` scatters the compacted diffusing rows back into a full-batch tensor before decoding, because the acoustic tokenizer's streaming `padding_cache` is indexed by batch row. Its output is therefore indexed by batch position, but the loop collecting the chunks read it by position among the diffusing rows.

The two agree only while every row emits an audio token on the same step. From the first step where they diverge -- in practice as soon as the shortest utterance in the batch emits EOS -- each remaining sequence starts collecting the previous row's audio, and the first one collects the zero-filled slot of a sequence that has already finished. The result is a clip that begins with its own speech and ends in silence or a neighbour's words, at the correct total duration: the model's own state uses the right indexing throughout, so only the saved waveform is misfiled and generation is otherwise unaffected.

Measured on 8 Seed-TTS English prompts with VibeVoice-1.5B: 41.3% WER at batch size 4 against 0.0% at batch size 1, and 0.0% at batch size 4 with this fix.

The added test generates a batch and the same sequences one at a time and compares them. Getting it to fail on the bug needs three things: rows that stop at different steps, an acoustic tokenizer whose output is actually distinguishable between rows (the tester defaults decode everything to ~1e-7, below the comparison tolerance), and fixed inputs -- `ids_tensor`'s module-level RNG is not seeded by `set_seed`.

### CODE REVIEW

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 1]
```diff
@@ -302,6 +296,83 @@ def test_vibevoice_generate_max_new_tokens(self):
...
+        # Guard against the test quietly becoming vacuous (see the docstring).
+        generated_lengths = {audio.shape[-1] for audio in batched.audio if audio is not None}
+        self.assertGreater(
+            len(generated_lengths),
+            1,
+            "This test only exercises the batched audio path if the sequences generate differing "
+            "amounts of audio; adjust the inputs until they do.",
+        )
+
+        self.assertEqual(len(batched.audio), input_ids.shape[0])
```
Comment (ebezzam) [inline tests/models/vibevoice/test_modeling_vibevoice.py]:
```suggestion
```

**Comment by ebezzam** [comment]
run-slow: vibevoice

File: src/transformers/models/vibevoice/generation_vibevoice.py [hunk 2]
```diff
@@ -688,8 +688,8 @@ def _sample(
                 )
                 audio_output = self._decode_audio_latent(audio_latent, diffusion_mask, batch_size, acoustic_cache)
                 acoustic_cache = audio_output.padding_cache
-                for i, sample_idx in enumerate(diffusion_mask.nonzero(as_tuple=False).view(-1)):
-                    audio_chunks[sample_idx.item()].append(audio_output.audio[i])
+                for sample_idx in diffusion_mask.nonzero(as_tuple=False).view(-1):
```
Comment (vasqu) [inline src/transformers/models/vibevoice/generation_vibevoice.py:691]: oh wow that is easy to miss tho tbf

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 3]
```diff
@@ -302,6 +296,73 @@ def test_vibevoice_generate_max_new_tokens(self):
         self.assertIsNotNone(output.audio)
         self.assertEqual(len(output.audio), self.model_tester.batch_size)
 
+    @pytest.mark.generate
+    def test_generate_batched_matches_single(self):
+        # different default to trigger error for incorrect indexing of audio chunks
```
Comment (vasqu) [inline tests/models/vibevoice/test_modeling_vibevoice.py]: Can you reference the issue / PR

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 4]
```diff
@@ -302,6 +296,73 @@ def test_vibevoice_generate_max_new_tokens(self):
...
+                "initializer_range": 0.5,
+                "weight_init_value": 0.5,
+            },
+        )
+        config = model_tester.get_config()
+        seed = 7
+        input_ids = ids_tensor(
+            [model_tester.batch_size, model_tester.seq_length], model_tester.vocab_size, rng=random.Random(seed)
+        )
+        attention_mask = torch.ones_like(input_ids)
```
Comment (vasqu) [inline tests/models/vibevoice/test_modeling_vibevoice.py]: Imo, we should just be able to use our usual create input and model helpers no? No need to construct a whole new tester etc

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 5]
```diff
@@ -302,6 +296,73 @@ def test_vibevoice_generate_max_new_tokens(self):
...
+            "return_dict_in_generate": True,
+            "guidance_scale": 1.3,
+            "num_diffusion_steps": 10,
+        }
+
+        # Initialize diffusion with same input for comparable outputs
+        def zeros_instead_of_randn(*args, **kwargs):
+            return torch.zeros(*args, **kwargs)
+
+        with torch.no_grad(), patch("torch.randn", zeros_instead_of_randn):
```
Comment (vasqu) [inline tests/models/vibevoice/test_modeling_vibevoice.py]: no grad should already be in generate no?

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 6]
```diff
@@ -302,6 +296,73 @@ def test_vibevoice_generate_max_new_tokens(self):
...
+                batched.audio[i] is None,
+                single.audio[0] is None,
+                msg=f"Sequence {i}: batched and single-sample generation disagree on whether audio was produced",
+            )
+            if batched.audio[i] is not None:
+                torch.testing.assert_close(
+                    batched.audio[i],
+                    single.audio[0],
+                    msg=lambda m, i=i: f"Sequence {i} differs between batched and single-sample generation:\n{m}",
+                )
```
Comment (vasqu) [inline tests/models/vibevoice/test_modeling_vibevoice.py:356]: lets also make the 2 split test like in older models pls (e.g. see mamba2)

**Review by vasqu — COMMENTED** [review]
Overall should be good just the test design should reuse our helpers and creators tbh

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 7]
```diff
@@ -302,6 +296,75 @@ def test_vibevoice_generate_max_new_tokens(self):
...
+
+    @pytest.mark.generate
+    def test_batched_equivalence_with_cache(self):
+        """Verifies that batched generation matches individual generation."""
+        self._check_batched_matches_single(use_cache=True)
+
+    @pytest.mark.generate
+    def test_batched_equivalence_without_cache(self):
+        """Verifies that batched generation matches individual generation, without cache."""
+        self._check_batched_matches_single(use_cache=False)
```
Comment (ebezzam) [inline tests/models/vibevoice/test_modeling_vibevoice.py:366]: @vasqu is this what you meant by 2 split test like [mamba2](https://github.com/huggingface/transformers/blob/3713bd839e580d07e4b70f2c89e986cb3c0e8ddf/tests/models/mamba2/test_modeling_mamba2.py#L469-L499)?

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 8]
```diff
@@ -302,6 +296,75 @@ def test_vibevoice_generate_max_new_tokens(self):
...
+        batch_size, seq_length = 4, 4
+
+        config = self.model_tester.get_config()
+        # Change config so the decoded audio is distinguishable between rows.
+        config.audio_config.layer_scale_init_value = 0.1
+        config.audio_config.initializer_range = 0.5
+        config.audio_config.weight_init_value = 0.5
+
+        input_ids = ids_tensor([batch_size, seq_length], self.model_tester.vocab_size, rng=random.Random(seed))
+        attention_mask = torch.ones_like(input_ids)
```
Comment (ebezzam) [inline tests/models/vibevoice/test_modeling_vibevoice.py:314]: re [your comment](https://github.com/huggingface/transformers/pull/48902#discussion_r4037817451), I'm not creating a whole new test this time, but are we go to overwrite some config parameters? so that the test would fail in the original case

Comment (vasqu) [reply tests/models/vibevoice/test_modeling_vibevoice.py:314]: Ah I think I confused you a bit / I worded it wrong: I just wanted the pattern as in https://github.com/ebezzam/transformers/blob/7b633379a4254ce92d599ae46e3e057810967837/tests/models/vibevoice/test_modeling_vibevoice.py#L231

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 9]
```diff
@@ -302,6 +296,75 @@ def test_vibevoice_generate_max_new_tokens(self):
...
+        seed = 7
+        batch_size, seq_length = 4, 4
+
+        config = self.model_tester.get_config()
+        # Change config so the decoded audio is distinguishable between rows.
+        config.audio_config.layer_scale_init_value = 0.1
+        config.audio_config.initializer_range = 0.5
+        config.audio_config.weight_init_value = 0.5
+
+        input_ids = ids_tensor([batch_size, seq_length], self.model_tester.vocab_size, rng=random.Random(seed))
```
Comment (vasqu) [inline tests/models/vibevoice/test_modeling_vibevoice.py:313]: Then you use create config and input as base passed to the test (in tester)

File: tests/models/vibevoice/test_modeling_vibevoice.py (hunk 7 above)
Comment (vasqu) [reply tests/models/vibevoice/test_modeling_vibevoice.py:366]: More like this https://github.com/huggingface/transformers/blob/3713bd839e580d07e4b70f2c89e986cb3c0e8ddf/tests/models/mamba2/test_modeling_mamba2.py#L319

But the design is good

**Review by vasqu — APPROVED** [review]
Last nits but overall good to go. Sorry i meant a different kind of split (i linked a bit more but on my phone so if its not clear just ping me again)

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 10]
```diff
@@ -155,10 +149,12 @@ def get_config(self):
             audio_token_id=5,  # Instead of default 151654
         )
 
-    def prepare_config_and_inputs(self):
-        config = self.get_config()
-        input_ids = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
-        attention_mask = torch.ones([self.batch_size, self.seq_length], dtype=torch.long, device=torch_device)
+    def prepare_config_and_inputs(self, batch_size=None, seq_length=None, rng=None, audio_config=None):
```
Comment (ebezzam) [inline tests/models/vibevoice/test_modeling_vibevoice.py]: ok to allow optional input args? such that they default to current behavior, but for the new test allow creating a bigger batch and different settings to trigger the error (for the original code)

File: tests/models/vibevoice/test_modeling_vibevoice.py [hunk 11]
```diff
@@ -302,6 +342,44 @@ def test_vibevoice_generate_max_new_tokens(self):
...
+            batch_size=4,
+            seq_length=4,
+            rng=random.Random(0),
+            audio_config={
+                **self.model_tester.audio_config,
+                "layer_scale_init_value": 0.1,
+                "initializer_range": 0.5,
+            },
+        )
+        self.model_tester.create_and_check_batched_matches_single(*config_and_inputs, use_cache=False)
```
Comment (ebezzam) [inline tests/models/vibevoice/test_modeling_vibevoice.py:381]: and so that we can do the two line pattern that I think you intended?

**Review by ebezzam — COMMENTED** [review]
@vasqu double checking if this is fine. from what I understood, you were looking to have a two-line pattern within the tests? 1. prepare inputs and 2. pass to tests

main question is if it's fine to add these optional args to the methods for preparing inputs? so we can keep the same small sizes for the other tests

File: tests/models/vibevoice/test_modeling_vibevoice.py (hunk 10 above)
Comment (vasqu) [reply tests/models/vibevoice/test_modeling_vibevoice.py]: I like the idea! I'd only remove the rng, is it really needed?

Comment (ebezzam) [reply tests/models/vibevoice/test_modeling_vibevoice.py]: nice! and no, not really needed. The original would sometimes pass (if it got lucky with the input), but we don't need it for the fix (from what I've observed over multiple runs). Removing it, launching slow tests, and merging 🙂

**Comment by ebezzam** [comment]
run-slow: vibevoice

File: tests/models/vibevoice/test_modeling_vibevoice.py (hunk 10 above)
Comment (ebezzam) [reply tests/models/vibevoice/test_modeling_vibevoice.py]: hmm actually seeding might be needed (got an error due to merge queue checks), as random input ids can lead to a flaky test:
- some of the audio in the batch don't generate audio
- the argmax for deciding to emit audio is near-tie, which can cause a diff between batched/single due to float noise

(FYI: obsevations from profiling with Claude over 70 seeds)

or is it better to use the `is_flaky` decorator?

Comment (vasqu) [reply tests/models/vibevoice/test_modeling_vibevoice.py]: hmm yea no worries then, it was just a if it works situation :D
